src/main/java/uk/gov/justice/hmpps/prison/service/reference/OffenceService.java

```diff
@@ -156,5 +163,32 @@ private PageImpl<OffenceDto> convertToPaginatedDto(final Page<Offence> pageOfOff
         return new PageImpl<>(offences, pageable, pageOfOffences.getTotalElements());
     }
 
+    @Transactional
+    public void linkOffencesToSchedules(final List<OffenceToScheduleMappingDto> offencesToSchedules) {
+        final var offenceIds = offencesToSchedules.stream().map(o -> new PK(o.getOffenceCode(), o.getStatuteCode())).toList();
+        final var offences = offenceRepository.findAllById(offenceIds);
+
+        final var offencesByCode = StreamSupport.stream(offences.spliterator(), true)
+            .collect(toMap(Offence::getCode, Function.identity()));
+
+        // Filter out any if the offence doesn't exist + filter out if the offence_indicator already exists (no duplicates)
+        final var offenceIndicators = offencesToSchedules.stream()
+            .filter(o -> offencesByCode.containsKey(o.getOffenceCode()))
+            .map(o -> OffenceIndicator.builder()
+                .offence(offencesByCode.get(o.getOffenceCode()))
+                .indicatorCode(o.getSchedule().getCode())
+                .build()
+            )
+            .filter(oi -> offenceIndicatorRepository.countByIndicatorCodeAndOffence_Code(oi.getIndicatorCode(), oi.getOffence().getCode()) == 0L)
```

Comment:
There is an `existsBy` in spring repositories you could use instead.

Reply:
Thanks @ldlharper, good spot - have made the change and pushed up again